### ISSUE TITLE
Per Token FlatAudience Option

### ISSUE THREAD

**Issue by davidmdm** [lestrrat-go/jwx#836]
**Contribution Guidelines**

**Abstract**
I would like to be able to sign tokens with Flat Audience on a per token basis.
It is currently  possible to set this behavior globally via `jwt.Settings`, however it is not possible to to turn toggle this behavior at the time of signing the JWT.

The need for this feature arises when a JWT producer serves different clients with different `aud` format expectations. 

**Describe the proposed solution/change**

Allow this marshaling option to be set either as a SignOption or directly on the token:

```go
token , _ := jwt.
  NewBuilder().
  Claim("aud", "value").
  Build()

token.SetFlatAudience(true)

// Token will have "aud":"value" and not ["value"] 
jwt.Sign(token, ...)

// Alternatively
jwt.Sign(token, jwt.WithFlattenAudience(true))
```

**Comment by lestrrat** [comment]
I'm not disagreeing with you, but

> The need for this feature arises when a JWT producer serves different clients with different aud format expectations.

There's _that_ need in the wild? dang :)

**Comment by lestrrat** [comment]
(note: this is me speaking aloud)

I suppose a per-object option could be created, which could be utilized for future enhancements.

One decision we have to make is how to store these options.
A bit flag is a natural fit for this sort of purpose because it encodes multiple options fairly compactly.
However, we would need to decide on how to work with the global settings.
Case in point, the following three cases work fine:

```go
// case 1: local option not set, global option not set
tok := ...
json.Marshal(tok) // default behavior, aud is a JSON array

// case 2: local option is set, global option not set
tok := ...
tok.SetOptions(jwt.Options() | jwt.FlattenAudience) // assume argument is a uint64 bit flag
json.Marshal(tok) // forced behavior, aud is a simple string, if it can be

// case 3: local option not set, global option is set
jwt.Settings(jwt.WithFlattenedAudience(true))
tok := ...
json.Marshal(tok) // forced behavior, aud is a simple string, if it can be
```

However, what happens if we want to **enable the global option**, but **explicitly disable the FlattenAudience behavior for a particular object**? There's no good way to do this using a bit flag because you cannot tell the difference between the default value and an explicit opt-out (i.e. case 3 shows that the per-object option is overridden by the global one, but how do you make the per-object option take precedence?)

One possible way to do this would be to change the semantics of the global option. We could make `jwt.Settings(jwt.WithFlattenedAudience(true))` to mean that "by default all token objects have the
per-object FlattenAudience flag turned on".

This way the default behavior is kept as-is,  and then given that maybe you can do

```go
tok.SetOptions(jwt.Options() & ~jwt.FlattenAudience)
```
to explicitly disable flattened audience behavior.

**Comment by lestrrat** [comment]
Possible API

```go
type TokenOptions uint64 // bit flag type
tok.Options() // return TokenOptions
tok.SetOptions(TokenOption) // set options

tok.Options().Enable(jwt.FlattenAudience)
tok.Options().Disable(jwt.FlattenAudience)
tok.Options().IsEnabled(jwt.FlattenAudience) // return bool
```

**Comment by lestrrat** [comment]
@davidmdm I haven't ran too many tests, but if you are willing to give it a whirl, I have https://github.com/lestrrat-go/jwx/tree/gh-836

**Comment by davidmdm** [comment]
To my mind, once a local token option is available, the global setting is worth deprecating. If I am controlling the token marshal behaviour via token options I don't need to set global options.

As a user, I wouldn't mind if you can't opt out of the global settings. I would avoid using them.

As for implantation I was simply going to extend "types.StringList" so you can change its marshalling behaviour.

I really like the token.SetOptions api though.

**Comment by davidmdm** [comment]
Sorry, I had commented before without taking your branch for a whirl. I just did and it was great. Solves my problem exactly!
I tried running it with global settings set to true and false, and the local override always took precedence. The API is great!

**Comment by lestrrat** [comment]
@davidmdm Cool. let me think/flesh out the details before implementing this in the main branch.

**Comment by lestrrat** [comment]
@davidmdm FYI it has been released with v2.0.7
